Add Caching of SSR public pages. Update default Cache-Control HTTP header

tdonohue opened the pull request:
## References
* Relates to #1954 and #1921

## Description
This PR adds basic, in-memory page caching of all *public* server-side rendered pages.  As most SSR requests are for public (anonymous access) pages, this can significantly increase the speed of the site.  

(However, it has no impact on the speed of SSR for authenticated content -- as it's not possible to cache this access-restricted content since it is much more dynamic & may include user-specific, sensitive information.)

## Instructions for Reviewers
List of changes in this PR:
* Adds a new `serverSide` section to the `cache` settings in `config.yml`:
  ```
  # In-memory cache(s) of server-side rendered pages. These caches will store the most recently accessed public pages.
  # Pages are automatically added/dropped from these caches based on how recently they have been used.
  # Restarting the app clears all page caches.
  # NOTE: To control the cache size, use the "max" setting. Keep in mind, individual cached pages are usually small (<100KB).
  # Enabling *both* caches will mean that a page may be cached twice, once in each cache (but may expire at different times via timeToLive).
  serverSide:
    # Set to true to see all cache hits/misses/refreshes in your console logs. Useful for debugging SSR caching issues.
    debug: false
    # When enabled (i.e. max > 0), known bots will be sent pages from a server side cache specific for bots.
    # (Keep in mind, bot detection cannot be guarranteed. It is possible some bots will bypass this cache.)
    botCache:
      # Maximum number of pages to cache for known bots. Set to zero (0) to disable server side caching for bots.
      # Default is 1000, which means the 1000 most recently accessed public pages will be cached.
      # As all pages are cached in server memory, increasing this value will increase memory needs.
      # Individual cached pages are usually small (<100KB), so max=1000 should only require ~100MB of memory.
      max: 1000
      # Amount of time after which cached pages are considered stale (in ms). After becoming stale, the cached
      # copy is automatically refreshed on the next request.
      # NOTE: For the bot cache, this setting may impact how quickly search engine bots will index new content on your site.
      # For example, setting this to one week may mean that search engine bots may not find all new content for one week.
      timeToLive: 86400000 # 1 day
      # When set to true, after timeToLive expires, the next request will receive the *cached* page & then re-render the page
      # behind the scenes to update the cache. This ensures users primarily interact with the cache, but may receive stale pages (older than timeToLive).
      # When set to false, after timeToLive expires, the next request will wait on SSR to complete & receive a fresh page (which is then saved to cache).
      # This ensures stale pages (older than timeToLive) are never returned from the cache, but some users will wait on SSR.
      allowStale: true
    # When enabled (i.e. max > 0), all anonymous users will be sent pages from a server side cache.
    # This allows anonymous users to interact more quickly with the site, but also means they may see slightly
    # outdated content (based on timeToLive)
    anonymousCache:
      # Maximum number of pages to cache. Default is zero (0) which means anonymous user cache is disabled.
      # As all pages are cached in server memory, increasing this value will increase memory needs.
      # Individual cached pages are usually small (<100KB), so a value of max=1000 would only require ~100MB of memory. 
      max: 0
      # Amount of time after which cached pages are considered stale (in ms). After becoming stale, the cached
      # copy is automatically refreshed on the next request.
      # NOTE: For the anonymous cache, it is recommended to keep this value low to avoid anonymous users seeing outdated content.
      timeToLive: 10000 # 10 seconds
      # When set to true, after timeToLive expires, the next request will receive the *cached* page & then re-render the page
      # behind the scenes to update the cache. This ensures users primarily interact with the cache, but may receive stale pages (older than timeToLive).
      # When set to false, after timeToLive expires, the next request will wait on SSR to complete & receive a fresh page (which is then saved to cache).
      # This ensures stale pages (older than timeToLive) are never returned from the cache, but some users will wait on SSR.
      allowStale: true
    ```
* Updates the `Cache-Control` HTTP Header behavior to help a user's browser to more appropriately cache files:
  * For static files (including *.js), default is now `Cache-Control: max-age=604800`, which sets it to one week.  This lets a user's browser cache these files for a week without needing to re-request them frequently (previously this was set to one *minute*).
      * Because we have `outputHashing: "all"` enabled in our [Angular build](https://angular.io/cli/build) settings in `angular.json`, all our compiled JS files automatically include a unique hash in their name.  This hash updates every time the content of the JS file changes.  Because the filename changes when updated, this will bypass the `Cache-Control` settings whenever these JS files are changed.
  * However, `index.html` now includes a `<meta>` directive to tell browsers not to cache it.  This is necessary for `outputHashing` to work right, as the `index.html` is updated with the new names of any newly compiled JS files.

HOW TO TEST:
* HINT: For easier cache testing, enable the `serverSide.debug: true` flag in your `config.*.yml`.  When enabled, all cache hits, misses & refreshes are logged.
* Build/deploy this PR. By default, ONLY the "bot" cache is enabled.
* First, as a normal anonymous user, verify that SSR is triggered on every request. Reload the page a few times, and verify you are seeing responses over over a second (at best).  If you'd enabled the "debug" logging, you'll also see no cache hits.
* Now, test that the "bot" cache is working. You can test it by temporarily changing your browser's user-agent to match that of a known bot. For example in Chrome:
    * Open up Dev Tools. Click on the three dots in upper right.
    * Go to "More Tools" and select "Network Conditions"
    * Deselect "Use browser default" for "User Agent".  In the dropdown, select "Googlebot".   This browser tab will now act like a Googlebot.
* Access the homepage. Verify you see a 200 OK response initially.  This first access will also take longer, as the server-side cache has not been initialized. On my local machine this first access usually takes ~2-3 seconds to load:
    ```
    GET /home 200 2220.162 ms - -
    ```
* Click reload in your browser. Normally, this would trigger SSR again.  However, with this PR you'll be accessing the page from cache and it'll load in just a few milliseconds.  If you enabled debugging, you'll also see a cache hit for the "bot cache":
    ```
    CACHE HIT FOR /home in bot cache
    GET /home 304 3.711 ms - -
    ```
* Now remove your "user-agent" so that you are no longer a "bot".
* Change your "config.prod.yml" settings to enable the "anonymous" cache & restart the UI:
  ```
  cache:
    serverSide:
      anonymousCache:
        max: 1000
  ```
* As an anonymous user, verify the cache is working.  The first request should take a second or two (and trigger SSR). If you reload the same page, you'll see it loads in a few milliseconds. If you enabled debugging, you'll also see a cache hit for the "anonymous cache":
  ```
  CACHE HIT FOR /home in anonymous cache
  GET /home 200 6.759 ms - 
  ```
    * NOTE: After the page is cached, you can also access the same page from a different local browser.  In that scenario, you'll get the cached version, but a 200 OK response (as it's your first access of that page).  This is because the cached copy is shared among all anonymous users.
* Verify that the anonymous cache reloads frequently. Wait for at least 10 seconds, and reload your page. You should see this in the logs:
   ```
   CACHE HIT FOR /home in anonymous cache
   CACHE EXPIRED FOR /home in anonymous cache. Re-rendering...
   GET /home 200 6.759 ms - -
   ```
* Verify there is no obvious differences in behavior after the cached copy of the page loads. (There shouldn't be, as the cached copy is quickly replaced by the CSR version.)
* Test a few other pages.  Initial page access will take longer, and the next reload will be very quick.
* Login. Verify that the same caching never triggers while authenticated.  When clicking the reload in your browser, all pages will return a 200 OK response and all will show a slightly slower load time as they are triggering SSR.

tdonohue commented:
@artlowel, @ybnd  and @atarix83 : This may be of high interest to you.  I'm still testing things out on my end, but this *very basic* caching of server-side rendered public pages provides a **massive speed boost** to DSpace 7.  So far, I have not found any side effects / major bugs.  It only works for **public** (anonymous access) pages, but those are the ones which most frequently trigger SSR.

Feedback/thoughts welcome.  I definitely would want more eyes on this before we consider moving it forward....but it seems like a possible simple way to increase the performance of SSR.

ybnd commented on server.ts:
```diff
@@ -201,62 +212,191 @@ export function app() {
+ * @param res current response
+ */
+function clientSideRender(req, res) {
+  res.render(indexHtml, {
```
Calling `render()` here means that we're attempting SSR, even though it's supposed to trigger direct CSR...

I suppose this was introduced to support changing the base path in https://github.com/DSpace/dspace-angular/pull/1642, but like this we don't really have a proper fallback in case SSR fails

tdonohue replied:
Yes, that's a good point.  I also was wondering about that as I refactored this code in `server.ts` but forgot to test it out.  As you correctly noted, I didn't write this `res.render()` code, I just refactored it into a separate method.  But, we should look for a fix to this CSR fallback behavior (in a separate ticket/PR).

ybnd replied:
Submitted an issue here: https://github.com/DSpace/dspace-angular/issues/2034

ybnd commented on server.ts:
```diff
@@ -201,62 +212,191 @@ export function app() {
+ * Enable server-side caching of pages rendered via SSR.
+ */
+function enableCache() {
+  if (cacheEnabled()) {
```
This looks a bit confusing, would be more clear if the method was called something like `initCache`

tdonohue replied:
Good point.  Will rename to `initCache()`

ybnd reviewed (commented):
This seems to work well!

Haven't been able to find any real problems, except that the cache takes a little while to kick in when a new URL is requested in a burst

* I've seen about 10 identical requests go through full SSR before any cache hits
* ...but only at ~10 requests/second
* Not something we should fix on the caching level since it's so easy to just burst random URLs, just FYI.

Also noticed a pre-existing issue with the "direct CSR fallback" that probably should be addressed in a separate issue.

ybnd reviewed (approved):
Pre-populating the cache would be nice, but this PR is already very useful.

I haven't found any other issues and there's little CPU/memory impact, clear win!

ybnd commented:
@artlowel mentioned he was also going to review this tomorrow

artlowel commented on server.ts:
```diff
@@ -201,62 +212,191 @@ export function app() {
+      res.locals.ssr = true;  // mark response as SSR (enables text compression)
+      if (debug) { console.log(`CACHE HIT FOR ${key}`); }
+      // return page from cache to user
+      res.send(cachedCopy);
+
+      // Check if cached copy is expired (in this sitution key will now be gone from cache)
+      if (!cache.has(key)) {
+        if (debug) { console.log(`CACHE EXPIRED FOR ${key} Re-rendering...`); }
+        // Update cached copy by rerendering server-side
+        // NOTE: Cached copy was already returned to user above. So, this re-render is just to prepare for next user.
+        serverSideRender(req, res, false);
+      }
```
Wouldn't this mean we can return cached copies that are older than 15 minutes? Perhaps it's best to rerender before returning the response

tdonohue replied:
@artlowel : Yes, it's possible that the cached copy could be older than 15 minutes... it's just guaranteed to be at least 15 mins old (based on what you set `timeToLive` to).

The intention of this behavior is to ensure the SSR is occurring _behind the scenes_ as much as possible.  So, that users rarely need to wait on the SSR processing.  In this example, the user who encountered the "stale" page does still see the cached copy, instead of that user having to wait for SSR to generate a new copy.  

Either approach can work here (and there are pros/cons to both approaches). But my goal was to see if we could find a way that users *rarely* had to wait on SSR.

artlowel reviewed (commented):
Thanks @tdonohue, it's a great idea!

The code looks good, and the performance increase is impressive.

There are some things that could be perceived as bugs though, so the question will be whether it's worth the trade off. 

For example, if you add a new top-level community that shows up in the first page on the homepage, in the first 15 minutes afterwards, users will still get the old cached version. If they move to the second page client-side, they'll see the last element of their first page appear again as the first element of the second page.

---
**Page 1 as a logged-in user:**
<img width="478" alt="Screenshot 2023-01-12 at 10 40 06" src="https://user-images.githubusercontent.com/1567693/212034219-eab1e987-6d8b-45e8-b628-aa590848e4b8.png">

---
**Page 1 as an anonymous user:**
<img width="430" alt="Screenshot 2023-01-12 at 10 40 14" src="https://user-images.githubusercontent.com/1567693/212034322-12ac7154-fe87-4c52-ad60-c50d2e1daf12.png">

---
**Page 2 as an anonymous user:**
<img width="448" alt="Screenshot 2023-01-12 at 10 40 19" src="https://user-images.githubusercontent.com/1567693/212034423-8390dd53-f0f1-44c0-8de1-1781ac18dafc.png">

---

It's not hard to think of other similar examples. 

Not necessarily a dealbreaker, but something to consider

tdonohue commented:
Capturing notes from today's discussion (for anyone not in the DevMtg):

As discussed in today's meeting, we are pausing this particular PR until some further investigation can be done, as there may be a better solution:
* @artlowel noted that it may be better to move the caching of anonymous requests to the *backend* (REST API), as that'd likely provide similar behavior (very quick responses) without some of the tradeoffs with this SSR caching approach. 
* The main tradeoff is that anonymous users can sometimes see outdated content if they hit the SSR cache at a point when the page's content has updated. This may be seen as a bug at times.  The main issue is the rxjs data store is *cached* as part of the page in the SSR cache, and there's not an obvious way to clear/refresh that rxjs data.
* Therefore, if the REST API can provide cached data for anonymous users, then the SSR process itself may be much quicker (based on some tests done by @ybnd , @artlowel reported that approximately 70% of the time taken to do SSR is simply waiting on the REST API response). This has the advantage that the REST API caching can be triggered to auto-update when objects change (or there could be a regular scheduled process to update the cache on a regular basis), avoiding some of the tradeoffs of this SSR caching approach.
* @ybnd and @artlowel will investigate REST API response caching for anonymous requests.  It seems possible, but needs some analysis/testing on whether it will work as well.  Hopefully we'll have more to discuss next week (Jan 19).

In the meantime, this PR will remain open & will **not** be merged.  We all do agree that the `Cache-Control` Http Header changes in this PR are **necessary**.  It's just the SSR caching which may be replaced by REST API response caching (should the two options work similarly).

mwoodiupui commented:
The business logic layer knows when cached data become invalid due to model changes.  Using this information, we can be much more responsive than by depending on timeouts, and avoid presenting stale data.

The simplest approach would be to clear the cache whenever a DSO is updated, which should be a fairly rare occurrence.  It might not cost much more to segment the cache by DSO type, or even to decache the specific object.  In any case, the thing to do is probably just to drop the stale cache entries and let normal lookup reload them when they are wanted.  Reloading on update can be wasted motion and unnecessary cache pressure if an object is not re-fetched fairly quickly after update.

If we take a simple/stupid approach, we should carefully instrument the cache so as to discover whether there is sufficient reason to do something more sophisticated later.

mwoodiupui commented:
I think that most cache misses will be generated by requests that are rendered client-side, so caching only on fetch will still tend to substantially reduce the time for initial page composition.  (I wonder how we can instrument to test this hypothesis.)

atarix83 commented:
@tdonohue @artlowel 

i'd like to share an alternative solution we can take into account. This is a  
https://itnext.io/incremental-static-regeneration-for-angular-42b0a8440e53

tdonohue commented:
@artlowel , @atarix83 and @ybnd :  I reviewed the link that @atarix83 shared (https://itnext.io/incremental-static-regeneration-for-angular-42b0a8440e53) and the solution there is quite similar to the solution I've built in this PR.  It's slightly more enhanced (in terms of options). But it's still just caching the entire HTML of the page & refreshing the cache on a timed basis (which this PR already does).  So, I'm not seeing any major differences in the approaches (unless I'm overlooking something).

However, a few things I **realized** though in revisiting this PR and re-testing it today:
1. @artlowel 's "bug" noted in [this comment](https://github.com/DSpace/dspace-angular/pull/2033#pullrequestreview-1245256101) already occurs on `main`.  If an anonymous user renders the homepage, and *then* you create a top level community (which appears on the 1st page), the anonymous user will see the same behavior that @artlowel notes.  They will not see that new top-level community until they reload the page (by clicking reload in the browser). If they click to page two, the last community on page one is duplicated to page two.  As far as I can see, this PR provides the same behavior as `main` in that scenario (unless I'm overlooking something... please test it though)
2. In reviewing the solution at https://itnext.io/incremental-static-regeneration-for-angular-42b0a8440e53, I realized their approach was to refresh the html cache *more frequently* (every 5 seconds).  I tried that with this PR by simply setting:
```
cache:
  serverSide:
    max: 100
    timeToLive: 5000  #5 seconds
```
This works just as well as the 15 minute caching scenario (speed increase is the same).  In fact, **it arguably works better** as the cached copy is only a few seconds out of date.

Based on this, I don't know if we should count this PR's approach "out" quite yet.   I know this PR is simple, but if we decrease to a 5 second cache (by default), it looks like we still get the same benefits and decrease the chance of an out-of-date page.  We obviously are triggering more SSR (with a 5 second timeToLive), but the server-side code is _still_ not under as much load as no cache at all (like on `main`).

mwoodiupui commented:
I have noted before that we have a tendency to cache or batch things for IMHO far too long, and I'm not aware of any research or testing to support such long lifetimes.  Often seconds will save very nearly as much work as minutes.

tdonohue commented:
Thanks @ybnd for the summary of what you've found!  @artlowel should pass along the feedback from today, but here's a brief summary:

1. I'll look at your SSR caching of bots-only approach.  I'd rather not abandon this "SSR caching of anonymous accesses" approach altogether though, as I'm concerned the "bots-only" approach may not have much impact for high load sites (maybe I'll be proven wrong though once this gets into Production).  Therefore, I recommend we look at providing both options, but making it configurable as follows:
    * Default setting will be the "bots-only caching" approach with a high TTL (your approach)
    * Provide an option to switch over to a "anonymous caching" approach. But, we'll add docs to note that you'd also want to then set a very low TTL (5-10 seconds)
    * I'll look at your code and see if I can update this PR quickly to add those two options.
2. We talked about getting #2067 updated so that it only triggers for bots.  That seems reasonable to also try for 7.5
3. Finally, we'll see if we can all test the new #2069 PR to see if that is stable enough to also add into 7.5

Thanks again for all the research into this performance issue!

ybnd commented:
> Provide an option to switch over to a "anonymous caching" approach

@tdonohue I think we should also consider making them independent of each other, so both can be active at the same time

tdonohue commented:
@wwelling : I'm not sure we've looked at that (at least I have not). I'm not against the idea though, so feel free to create a PR if you wish.  I'm not sure that it's directly related to the issue that this PR tries to fix (that SSR is just really slow & easily overwhelmed for highly active sites), so we may want to treat that as a separate performance enhancement.  It still sounds like it would potentially provide a performance improvement here...so it would be good to look at how enabling it would impact DSpace.

tdonohue commented:
@ybnd and @artlowel : I've updated this PR to have two caches:
1. `bot` cache : enabled by default for 1,000 pages.  Pages expire after 1 day (by default).  Both settings are configurable.
    * I chose a 1 day expiration as I think this cache may impact SEO.  I expect most sites would want new items to appear in Google relatively quickly, so a day seems reasonable.
2. `anonymous` cache: disabled by default. But, when enabled, pages expire after 10 seconds (by default).  Both settings are configurable.
    * Per previous comments, I set the expiration very low so that, when enabled, anonymous users should not encounter outdated content frequently.

A few more notes:
* I've updated the description at the top with detailed instructions on how to test both caches, especially the "bot" one which requires changing your "user-agent" to a known bot string.
* These caches are entirely separate caches as they will need different `timeToLive` values.  This means that if you enable both, a single page can be stored in both caches (i.e. stored twice in memory) but expire at different times based on cache `timeToLive`. For example, once a page appears in the "bot" cache, it won't be updated until the next day... but the same page will keep updating every ~10 seconds in the "anonymous" cache if it continues to be accessed.
* When both caches are enabled, if a page triggers SSR the results will then be saved to *both caches* (if it doesn't exist in the cache or is expired in the cache).  This means that a request from a bot can update the "anonymous" cache, and a request from an anonymous user can update the "bot" cache. This is on purpose, as it minimizes SSR requests when both caches are enabled.
    * However, if SSR ever generates a *different* page for bots than for anonymous users, we'd obviously want to change this behavior. But, this behavior is currently safe as anonymous users and bots will both receive the same page from SSR.

tdonohue commented on server.ts:
```diff
@@ -215,62 +230,248 @@ export function app() {
+    if (!cache.has(key)) {
+      if (debug) { console.log(`CACHE EXPIRED FOR ${key} in ${cacheName} cache. Re-rendering...`); }
+      // Update cached copy by rerendering server-side
+      // NOTE: In this scenario the currently cached copy will be returned to the current user.
```
@ybnd : The reason I went with this approach is to attempt to perform as much SSR *behind the scenes* as possible.... i.e. this ensures bots/anonymous users are mostly interacting with the cache & not having to wait on SSR.

I do agree there is a risk here that, if a particular page isn't accessed for a long time after the `timeToLive` has passed, you'll receive a more out-of-date copy than expected.  But, since bots tend to access sites frequently and they will always trigger both caches to update, I felt this risk is rather minimal.

I do agree though that this implementation means `timeToLive` is more of a "minimum staleness" than a "maximum staleness".... But, I feel the gap between those definitions should be minimal for most sites, simply because bots will trigger the cache refresh on a regular basis, even if users are not highly active on the site.

ybnd replied:
Such a config option would be great!

I'm happy if it's `allowStale: true` by default; performance would indeed benefit, sites can still opt _and_ the potential trade-off is well-documented 👍

tdonohue commented on server.ts:
```diff
@@ -215,62 +230,248 @@ export function app() {
+ * @returns cached copy (if found) or undefined (if not found)
+ */
+function checkCacheForRequest(cacheName: string, cache: LRU<string, any>, req, res): any {
+  let debug = false; // Enable to see cache hits & re-rendering in logs
```
Makes sense to change this into a config property to allow sites to play with this more easily & allow for logging `CACHE MISS` (as you noted in your next comment).

tdonohue commented:
@ybnd : I've just pushed updates to add a `debug` config and the `allowStale` option (for both caches).  I've updated the PR description with those details.

ybnd commented on server.ts:
```diff
@@ -324,7 +324,7 @@ function initCache() {
       max: environment.cache.serverSide.botCache.max,
       ttl: environment.cache.serverSide.botCache.timeToLive || 24 * 60 * 60 * 1000, // 1 day
-      allowStale: true // If object is found to be stale, return stale value before deleting
+      allowStale: environment.cache.serverSide.botCache.allowStale || true // if object is stale, return stale value before deleting
```
Using the `||` operator here makes it so `allowStale` can never be turned off through configuration.
It should be a [`??`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) instead.

tdonohue replied:
@ybnd : Good catch!  I completely missed that ... and this was definitely a copy & paste error.  Just pushed up a fix a moment ago in https://github.com/DSpace/dspace-angular/pull/2033/commits/9e11d69a8ade6fae07676d49f6a2cb4697705107

artlowel reviewed (approved):
Thanks @tdonohue!

The code looks good, and it works well.

When @ybnd's issue with `allowStale` is fixed, this is a +1 for me

tdonohue commented:
Merging for 7.5.  I'll add documentation to https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Configuration
